src/main/java/com/databricks/jdbc/api/impl/DatabricksPreparedStatement.java

```
@@ -147,68 +147,254 @@ private boolean canUseBatchedInsert() {
     }
   }
 
+  /**
+   * Executes a chunk with retry logic. If it fails after retries, splits the chunk in half and
+   * tries each half separately (recursively).
+   *
+   * @param insertInfo Parsed INSERT statement information
+   * @param startIndex Starting row index (inclusive)
+   * @param endIndex Ending row index (exclusive)
+   * @param allUpdateCounts Array to store update counts
+   * @param interpolateParameters Whether to interpolate parameters into SQL
+   * @param depth Current recursion depth (0 = top level)
+   * @return Number of rows successfully processed
+   * @throws DatabricksBatchUpdateException if a single row fails after all retries
+   */
+  private int executeChunkWithRetryAndSplit(
```

I am only worried about this causing duplicate insertions without idempotency!

---

What I have observed is that the databricks API will reject a payload outright.
I have not seen any "Partial writes" - I have done some testing which triggers this logic explicitly (by using overly large batch sizes).

The validation query I run (using tpch datasets) is 
```
SELECT COUNT(*), COUNT(DISTINCT *) FROM <catalog>.<table>
```
I have done this with multiple spark executors and have not seen duplicates generated.

We can consider making the retries here less opinionated (ie. let the batches fail if the BATCH_SIZE is too big)
I am certain we should have some retries to account for server size back-pressure (rate limiting etc)..

---

Ack on the refactor. Will take care of that.

---

@samikshya-db - Following up on the duplicate data concern. 

Multi-row INSERT statements should be atomic in Databricks. A statement like this should insert all 3 rows or none:
```sql
INSERT INTO users VALUES (1, 'Alice'), (2, 'Bob'), (3, 'Charlie')
```

The retry logic only retries failed statements (which should have committed zero rows). Successful statements are never retried. Example with 10 rows where row 5 fails:
```
INSERT rows 0-9 (×4) → FAIL (0 rows committed)
  ├─ INSERT rows 0-4 → ✅ 5 committed, done
  └─ INSERT rows 5-9 → FAIL (0 committed)
      ├─ INSERT rows 5-6 → FAIL (0 committed)
      │   ├─ INSERT row 5 → ❌ FAIL
      │   └─ INSERT row 6 → ✅ 1 committed, done
      └─ INSERT rows 7-9 → ✅ 3 committed, done

Result: 9 rows inserted
```

Assuming INSERT atomicity, we only retry statements that were rejected by the server, so duplicates shouldn't occur. Is there's a scenario where Databricks SQL can partially succeed?

The prior duplicate data bug was related to a missing `clearBatch` call after a successful execute batch statement, which violated the JDBC spec. 

Please let me know if you have any questions.

---

logic removed.